### DISCUSSION TITLE
Metrics

### DISCUSSION THREAD

**Discussion by estahn** [alitto/pond#8] [Ideas]
I would like to expose some metrics about the state of the pool in my application, e.g. the number of items in the backlog, the number of workers, etc.

Thoughts?

**Comment by alitto** [comment]
Hey @estahn!

So far, only 2 metrics are exposed by pools:
- `pool.Running()`: returns the number of active/running worker goroutines (gauge)
- `pool.Idle()`: returns the number of idle worker goroutines (gauge)

The number of worker goroutines currently executing a task can be calculated as `pool.Running() - pool.Idle()`.

It would be very useful to expose more metrics, such as:
- `pool.MaxWorkers()`: Max number of workers (constant/gauge) 
- `pool.MinWorkers()`: Min number of workers (constant/gauge)
- `pool.SubmittedCount()`: Number of tasks submitted since the pool was created (counter)

If you think of any other metric that could be useful please let me know. I might send a PR to add these soon.

**Comment by alitto** [comment]
FYI @estahn, v1.5.0 is out and it adds a number of useful metrics to worker pools, see https://github.com/alitto/pond#metrics--monitoring for more info. I've also added an example to show how these can be exposed to [Prometheus](https://github.com/alitto/pond/blob/master/examples/prometheus/prometheus.go)
Cheers

**Reply by estahn** [reply]
@alitto Awesome, thanks. I will check it out soon. I'm using your lib here btw: https://github.com/estahn/k8s-image-swapper

**Reply by alitto** [reply]
That's great man! nice tool, certainly very useful now that Docker hub rate limits pulls 🙂
